### ISSUE TITLE
CDM Database connection information outside Source table

### ISSUE THREAD

**Issue by davidhcar** [OHDSI/WebAPI#2227]
We have a security concern storing password in the source table that holds database connection information. Wondering if this is the only way to store, Are there Settings.xml configuration variables used in place of Source table in DB? looking at the sample settings.xml does not seem to facilitate this, but wanted to get the confirmation. Could someone advise on this please ?

**Comment by chrisknoll** [comment]
Yes, it is the only way to store, but the password is encrypted.

**Comment by davidhcar** [comment]
Thank you @chrisknoll for the prompt response. Could you please add more information/context around how it is encrypted, and how the encryption works ?

**Comment by chrisknoll** [comment]
Sure.

The bean that handles encryption is defined [here](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/DataAccessConfig.java#L116).  Please see documentation  for details about the PBESringEncryptor from the jasypt encryption library.

This bean is injected into the SourceSource service via constructor [here](https://github.com/OHDSI/WebAPI/blob/master/src/main/java/org/ohdsi/webapi/source/SourceService.java#L40).

The way it works is, if you have the [jasypt.encryptor.enabled](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/source/SourceService.java#L26) flag set in your application settings, when [ensureSourceEncrypted()](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/source/SourceService.java#L54) is called, it will fetch all the `SOURCE` records from the webapi database, and update the passwords to an [encoded value](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/source/SourceService.java#L68). 

The `ensureSourceEncrypted()` method is called when the application starts up (via a `@PostConstruct [directive](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/source/SourceService.java#L48)), or when the `/refresh` endpoint is [invoked](https://github.com/OHDSI/WebAPI/blob/fe1920b1d4173a4e5cd4e3c023b9c53aeeba3ed5/src/main/java/org/ohdsi/webapi/source/SourceController.java#L89). 

You can tell if your own passwords are encrypted by looking at your password column, and seeing if it starts with `ENC(`.

**Comment by davidhcar** [comment]
Thank you @chrisknoll for the descriptive response, I believe these properties can be added in the settings.xml. Couple more questions to clarify,
1. `jasypt.encryptor.algorithm` - Is there a list of supported algorithms somewhere, we might need to update the default one with more advanced depending on our security team advise.
2. What is the use of `jasypt.encryptor.password` when we store the password in the database table.

   ```
 <!-- PASSWORD ENCRYPTION -->
    <jasypt.encryptor.enabled>false</jasypt.encryptor.enabled>
    <jasypt.encryptor.password></jasypt.encryptor.password>
    <jasypt.encryptor.algorithm>PBEWithMD5AndDES</jasypt.encryptor.algorithm>
```

**Comment by chrisknoll** [comment]
I've tried to find references to those other configuration parameters in the code, but I can't find where they are referenced (other than `jasypt.encryptor.enabled`).  I'd have to defer to the folks who worked on [this PR](https://github.com/OHDSI/WebAPI/pull/390).

@wivern , can you explain what the password and algorithm params of jasypt.encrytor do?

From my perspective, I believe all the application properties are exported to an Environment which is a Spring feature that defines the 'runtime context'.   So even tho I can't find specific references to these variables, it could be that they are defined in an Environment and is usable by the jasypt framework.   My understanding of jsypt is that it doesn't perform the encryption but allows the lookup of encryption providers.

I'm not sure why we exposed paramaters for this, it very well probably could have worked that we use jasypt 'under the covers' and that the encryption method is not configurable.  But I think the idea was to allow it to be configurable, but there's no documentation to describe it (that I'm aware of).

**Comment by davidhcar** [comment]
Given the weakness with 'PBEWithMD5AndDES' algorithm, we may not be able to use this. If there is a ENUM list that would help to determine stronger algorithm to use in the settings.

**Comment by davidhcar** [comment]
Thank you very much @chrisknoll for digging into this. @wivern @anton-abushkevich Appreciate your insights as well.

**Comment by davidhcar** [comment]
Enabling the` jasypt.encryptor.enabled = true` property causes application startup failed with following errors, appreciate your insights on this. Source table has password value.
```

Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'ccController' defined in file [/usr/local/tomcat/webapps/WebAPI/WEB-INF/classes/org/ohdsi/webapi/cohortcharacterization/CcController.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'ccServiceImpl' defined in file [/usr/local/tomcat/webapps/WebAPI/WEB-INF/classes/org/ohdsi/webapi/cohortcharacterization/CcServiceImpl.class]: Unsatisfied dependency expressed through constructor parameter 4; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'feAnalysisServiceImpl' defined in file [/usr/local/tomcat/webapps/WebAPI/WEB-INF/classes/org/ohdsi/webapi/feanalysis/FeAnalysisServiceImpl.class]: Unsatisfied dependency expressed through constructor parameter 3; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'vocabularyService': Unsatisfied dependency expressed through field 'sourceService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sourceService': Invocation of init method failed; nested exception is org.jasypt.exceptions.EncryptionInitializationException: **Password not set for Password Based Encryptor**
2023-04-06 18:04:37.219 INFO main org.springframework.scheduling.concurrent.ExecutorConfigurationSupport - [] - Shutting down ExecutorService 'taskExecutor'
2023-04-06 18:04:38.090 INFO main org.springframework.boot.autoconfigure.logging.AutoConfigurationReportLoggingInitializer - [] -
2023-04-06 18:04:38.196 ERROR main org.springframework.boot.SpringApplication - [] - Application startup failed
```
